NPE in multiple define commands in one statement after upgrade from H2 4.1.197

We upgraded H2 from 1.4.197 to 1.4.200 and we are now hitting a NPE. I narrowed down the case and it seems to be related to executing multiple statements in 1, with a specific sequence. Out of curiosity I tried 1.4.198 and 1.4.199 and the bug also occurs there.

The statements (to be executed in a single statement):
CREATE TABLE TblA (AField DOUBLE);
CREATE INDEX TblAIndex ON TblA(AField);
ALTER TABLE TblA ADD (PrevDate DOUBLE);
CREATE TABLE TblB(AField DOUBLE) AS SELECT AField FROM TblA;

The NPE (MV_STORE FALSE):
Caused by: java.lang.NullPointerException
at com.h2database/org.h2.pagestore.db.PageStoreTable.getRowCountApproximation(PageStoreTable.java:510)
at com.h2database/org.h2.pagestore.db.ScanIndex.getCost(ScanIndex.java:122)
at com.h2database/org.h2.table.TableFilter.getBestPlanItem(TableFilter.java:234)
at com.h2database/org.h2.table.Plan.calculateCost(Plan.java:120)
at com.h2database/org.h2.command.dml.Optimizer.testPlan(Optimizer.java:181)

The NPE (MV_STORE TRUE):
Caused by: java.lang.NullPointerException
at com.h2database/org.h2.mvstore.db.MVTable.lock(MVTable.java:132)
at com.h2database/org.h2.table.TableFilter.lock(TableFilter.java:207)
at com.h2database/org.h2.command.dml.Select.queryWithoutCache(Select.java:819)
at com.h2database/org.h2.command.dml.Query.queryWithoutCacheLazyCheck(Query.java:201)
at com.h2database/org.h2.command.dml.Query.query(Query.java:489)
at com.h2database/org.h2.command.dml.Insert.insertRows(Insert.java:217)
at com.h2database/org.h2.command.dml.Insert.update(Insert.java:151)
at com.h2database/org.h2.command.ddl.CreateTable.update(CreateTable.java:145)

Here is a minimal test case which you can simply run. You can also easily switch between the 2 stores too.
[H2NpeTest.zip](https://github.com/h2database/h2database/files/3910176/H2NpeTest.zip)

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #2304: NPE in multiple define commands in one statement after upgrade from H2 4.1.197
+</li>
 <li>PR #2303: Assorted minor changes
 </li>
 <li>Issue #2286: Inline check constraints not in INFORMATION_SCHEMA

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -772,7 +772,7 @@ public class Parser {
             if (currentTokenType == SEMICOLON) {
                 String remaining = originalSQL.substring(parseIndex);
                 if (!StringUtils.isWhitespaceOrEmpty(remaining)) {
-                    return prepareCommandList(c, sql, remaining);
+                    return prepareCommandList(c, p, sql, remaining);
                 }
             }
             return c;
@@ -781,17 +781,16 @@ public class Parser {
         }
     }
 
-    private CommandList prepareCommandList(CommandContainer command, String sql, String remaining) {
+    private CommandList prepareCommandList(CommandContainer command, Prepared p, String sql, String remaining) {
         try {
             ArrayList<Prepared> list = Utils.newSmallArrayList();
-            boolean stop = false;
             do {
-                if (stop) {
+                if (p instanceof DefineCommand) {
+                    // Next commands may depend on results of this command.
                     return new CommandList(session, sql, command, list, parameters, remaining);
                 }
                 suppliedParameters = parameters;
                 suppliedParameterList = indexedParameterList;
-                Prepared p;
                 try {
                     p = parse(remaining);
                 } catch (DbException ex) {
@@ -800,10 +799,6 @@ public class Parser {
                         throw ex;
                     }
                     return new CommandList(session, sql, command, list, parameters, remaining);
-                }
-                if (p instanceof DefineCommand) {
-                    // Next commands may depend on results of this command.
-                    stop = true;
                 }
                 list.add(p);
                 if (currentTokenType == END) {

--- a/h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java
+++ b/h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java
@@ -1758,6 +1758,12 @@ public class TestPreparedStatement extends TestDb {
             assertFalse(rs.next());
         }
         stmt.execute("DROP TABLE TEST");
+        ps = conn.prepareStatement("CREATE TABLE A (C1 INT);" //
+                + "CREATE INDEX A_IDX ON A(C1);" //
+                + "ALTER TABLE A ADD (C2 INT);" //
+                + "CREATE TABLE B AS (SELECT C1 FROM A);");
+        ps.executeUpdate();
+        stmt.execute("DROP TABLE A, B");
     }
 
     private void testAfterRollback(Connection conn) throws SQLException {

